test/hashtree_eqc.erl

```diff
@@ -68,109 +114,265 @@ object() ->
 
-command(_S = #state{started = Started}) ->
+%% Generate tree ids - the first one is used for the test, the others are added
+%% as empty hashtrees.
+%%
+%% Make sure the TreeId is unique (the second element), does not matter if there
+%% are dupes in the first number (Index)
+ids() ->
+    ?SUCHTHAT(TreeIds,
+              non_empty(list({?LET(X, nat(), 1+X), ?LET(X,nat(), min(X,255))})),
+              %% not lists:member(hd(TreeIds), tl(TreeIds)) andalso
+              lists:keyfind(element(2, hd(TreeIds)), 2, tl(TreeIds)) == false).
+
+%%
+%% Generate the commands, split into two cases to force the start to happen as
+%% the first command.
+%%
+command(_S = #state{started = false}) ->
+    {call, ?MODULE, start, [params(), ids(), mark(), mark()]};
+command(_S = #state{started = true, tree_id = TreeId,
+                    params = {_Segments, _Width, MemLevels},
+                    snap1 = Snap1, snap2 = Snap2}) ->
+    %% Weights to increase snap frequency once update snapshot has begun
+    SS = Snap1 /= undefined orelse Snap2 /= undefined,
+    SF = case SS of true -> 100; _ -> 1 end,
     frequency(
-        [{1, {call, ?MODULE, start,
-              [{oneof(?POWERS), oneof(?POWERS), choose(0, 4)}, mark(), mark()]}} || not Started] ++
-        [{1, {call, ?MODULE, update_tree, [t1]}} || Started] ++
-        [{1, {call, ?MODULE, update_tree, [t2]}} || Started] ++
-        [{50, {call, ?MODULE, write, [t1, objects()]}} || Started] ++
-        [{50, {call, ?MODULE, write, [t2, objects()]}} || Started] ++
-        [{100,{call, ?MODULE, write_both, [objects()]}} || Started] ++
-        [{50, {call, ?MODULE, delete, [t1, key()]}} || Started] ++
-        [{50, {call, ?MODULE, delete, [t2, key()]}} || Started] ++
-        [{100,{call, ?MODULE, delete_both, [key()]}} || Started] ++
-        [{1, {call, ?MODULE, reopen_tree, [t1]}} || Started] ++
-        [{1, {call, ?MODULE, reopen_tree, [t2]}} || Started] ++
-        [{1, {call, ?MODULE, unsafe_close, [t1]}} || Started] ++
-        [{1, {call, ?MODULE, unsafe_close, [t2]}} || Started] ++
-        [{1, {call, ?MODULE, rehash_tree, [t1]}} || Started] ++
-        [{1, {call, ?MODULE, rehash_tree, [t2]}} || Started] ++
-        [{1, {call, ?MODULE, reconcile, []}} || Started] ++
-        []
+      %% Update snapshots/trees. If memory is enabled must test with update_tree
+      %% If not, can use the method used by kv/yz_index_hashtree and separate
+      %% the two steps, dumping the result from update_perform.
+        [{10*SF, {call, ?MODULE, update_tree,  [t1, s1]}} || Snap1 == undefined] ++
+	[{10*SF, {call, ?MODULE, update_snapshot,  [t1, s1]}} || Snap1 == undefined, MemLevels == 0] ++
+	[{10*SF, {call, ?MODULE, update_perform,   [t1]}} || Snap1 == created, MemLevels == 0] ++
+	[{10*SF, {call, ?MODULE, set_next_rebuild, [t1]}} || Snap1 == updated] ++
+        [{10*SF, {call, ?MODULE, update_tree,  [t2, s2]}} || Snap1 == undefined] ++
```

Comment:
Should this check be `|| Snap2 == undefined`? I haven't actually worked through all the implications of whether this makes sense, but my gut feeling is that this might be a typo, just based on it not matching up with the pattern of how all the other calls in this list are structured.